### ISSUE TITLE
Web GUI Framework Design

### ISSUE THREAD

**Issue by Supernova1114** [SaddlebackCollegeRobotics/.github#135]
A Web GUI seems like a more viable solution than a normal software interface:
- Interface can be used on any system.
- Can easily use a mobile device to send commands to the rover.
- With the web-server hosted on the rover, we would not need to have any extra code on devices that need to communicate with it.
- Can be accessed through a browser.

ToDo:
- [x] Create list of the software stack and framework we will be using.

**Comment by devloos** [comment]
### Documentation
- Our web client communicates with our Python server via the HTTP protocol and through web sockets using Django channels.

### Demo
https://github.com/SaddlebackCollegeRobotics/.github/assets/58356571/1ba2346d-b29d-4bd6-8694-4b1735a9df58

### Todos
- figure out live video
- figure out how to get a console that's connected to our base station terminal
- figure out how to handle config files for the rover

### Resources
- [Web Socket Tutorial](https://www.youtube.com/watch?v=cw8-KFVXpTE)
- [Django Python Folder Structure](https://medium.com/django-unleashed/django-project-structure-a-comprehensive-guide-4b2ddbf2b6b8)

**Comment by devloos** [comment]
- Video feed of the rover: from the zed
- Watchdog status: is he starving or not
- Command console
- GPS map
- Live Lidar scans
- Antenna signal strength
- Different feature setups/layouts for different missions

**Comment by devloos** [comment]
General:

-  Battery level
- Drive speed
- Current
- Longitude and latitude 
- Heading 
- Distance traveled
- Military time clock
- Antenna signal strength

Arm:

- Visual model of the current position of the arm and its axis
- Retract the arm

Science:

- Sensor data
- Sensor visualization, i.e graphs

Autonomy:

- Bounding box for detected object
- Input for GPS coordinates
- Map height and waypoint visualization
- Rover position and ability to add points to map by clicking
- Box with a colour that is the same as the one on the rover that flashes when the rover arrives at a destination, drives autonomously, etc...
- Green pathfinding line from ROS navigation stack

**Comment by devloos** [comment]
- Video feed of the rover
  - Feed from zed
  - Arm cameras
- Watchdog status
  - We will probably need 10
  - Should we have the ability to reset?
- Command console
  - is there a library
  - lets restrict some commands
- GPS map
  - something like boat GPS system
  - ![Image](https://github.com/SaddlebackCollegeRobotics/.github/assets/58356571/8fb0d0f5-92ba-4ea5-85f4-6830d0455789)
- Live Lidar scans
  - can I integrate this with gps
  - I should search for a library
- Different feature setups/layouts for different missions

**Comment by devloos** [comment]
Terminal libraries offering complete SSH capabilities cannot be seamlessly integrated into another project. In simpler terms, such libraries function as standalone projects and do not integrate well with our existing projects.

While it is possible to implement a basic shell for executing and transmitting commands, there are limitations. This approach only supports basic commands without the ability to edit or navigate through the folder structure.

**Comment by devloos** [comment]
Foxglove:

It seems that the most effective way to establish a real-time connection is through the utilization of web sockets. The system supports diverse layouts, including the incorporation of tabs within each layout.

Key features for our needs:
Video feeds
Map display
Lidar visualization
Various metrics visualization options, such as watchdog status, antenna signal strength, drive speed, and more (not limited to heading)
Service calls
Custom message schemas

Limitations to consider:
Inability to perform action calls
Incapacity to read or set ROS parameters

The implementation process is relatively straightforward but necessitates an additional node, namely 'foxglove_bridge', to be executed. This particular node actively listens to topics being published by other nodes. While certain node settings can be adjusted to modify behavior, ultimately, the data is transmitted through a web socket.

Despite these intricacies, setting up a video feed and visualizing the robot took only a few minutes to accomplish.

Also the video feed you are seeing is compressed and sent via a standard ros2 sensor msg "CompressedImage"

Foxglove understands most if not all std msgs out the gate
